### PR TITLE
Initial sequence diagrams

### PR DESCRIPTION
**Pull request by danielburnley**
Adding three sequence diagrams in mermaid:
- Filling in a form
- Creating and publishing a form
- Updating a published form

### CODE REVIEW

File: diagrams/sequence-diagrams/creating-and-publishing-a-form.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+```mermaid 
+sequenceDiagram
+  actor user as User
+  participant manager as Manager
+  participant designer as Designer
+  participant api as API
+
+  user -->> manager: Creates a new form
+  manager -->> api: Submit form creation information
+  api -->> manager: Returns new form information
```
Comment (james-sheppard-gov-uk) [inline diagrams/sequence-diagrams/creating-and-publishing-a-form.md:12]: What new form information is being returned? Won't it just be `success` or `error`?

File: diagrams/sequence-diagrams/creating-and-publishing-a-form.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  manager -->> user: Redirects user to designer for new form
+  loop form creation
+    user -->> designer: Updates form
+    designer -->> api: save updated form
+    api -->> api: verify and process form
+    api -->> designer: return updated form
```
Comment (james-sheppard-gov-uk) [inline diagrams/sequence-diagrams/creating-and-publishing-a-form.md:18]: What "updated form" is being returned? Won't it just be a `success` or `error` response?

File: diagrams/sequence-diagrams/creating-and-publishing-a-form.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    designer -->> user: progress to next screen
```
Comment (james-sheppard-gov-uk) [inline diagrams/sequence-diagrams/creating-and-publishing-a-form.md:19]: Will the user always progress to a new screen? What about in-screen edits?

File: diagrams/sequence-diagrams/creating-and-publishing-a-form.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Creating and publishing a form
+
+```mermaid 
+sequenceDiagram
+  actor user as User
+  participant manager as Manager
+  participant designer as Designer
+  participant api as API
+
+  user -->> manager: Creates a new form
```
Comment (james-sheppard-gov-uk) [inline diagrams/sequence-diagrams/creating-and-publishing-a-form.md:10]: We may want to reference an existing form (e.g. page on GOV.UK) that this is the online version of.

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+sequenceDiagram
+  actor user as User
+  participant runner as Runner
+  participant api as API
+  participant store as User data store
+  participant submitter as Submitter
+  participant notify as GOV.UK Notify
+  actor processor as Processor
+
+  user -->> runner: Navigates to form from GOV.UK
```
Comment (james-sheppard-gov-uk) [inline diagrams/sequence-diagrams/filling-in-a-form.md:13]: We should be more specific about where the form "start page" lives

File: diagrams/sequence-diagrams/creating-and-publishing-a-form.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    api -->> designer: return updated form
+    designer -->> user: progress to next screen
+  end
+
+  note over user,api: Potential for the review process to go here
+
+  user -->> manager: publishes form
+  manager -->> api: marks form as published
+  api -->> manager: returns published form URL
+  manager -->> user: displays page with published url
```
Comment (danielburnley) [inline diagrams/sequence-diagrams/creating-and-publishing-a-form.md:27]: Add something for putting the start page on GOV.UK?

**Comment by james-sheppard-gov-uk** [comment]
I've added a README.md and individual files for
* authentication
* creating a form
* publishing a form
* filling in a form
* changing a form

There are still a few TODOs but comments very welcome on accuracy / completeness so far...

**Comment by james-sheppard-gov-uk** [comment]
Agreed in "Weekly forms technical catch up" today that we should get this merged ASAP then incrementally work to improve it.
* changing a form

is still a `TODO` and should be considered with the draft/live work currently underway.

**Comment by danworth** [comment]
Just wondering if these diagrams would be better on the https://github.com/alphagov/forms-team/wiki? There's no PR flow for the wiki which is either a good thing by reducing the barrier to contributing or a bad thing because its harder to propose a change.

File: diagrams/sequence-diagrams/creating-a-form.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+  link api: GitHub repo @ https://github.com/alphagov/forms-api
+
+  note right of user: user has already logged in
+
+  user->>browser: click "Create a form >" button
+  browser->>admin: GET /forms/new
+  browser-->>user: show "What is the name of your form?" page
+  user->>browser: Give the form a name<br />click "Save and continue" button
+  browser->>admin: POST /forms/new<br />payload: forms_change_name_form%5Bname%5D={form name}
+  admin->>api: POST /forms<br />{"name":"{form name}",<br />"submission_email":"",<br />"org":"{user orginisation}"}
```
Comment (aliuk2012) [inline diagrams/sequence-diagrams/creating-a-form.md:20]:
```suggestion
  admin->>api: POST /forms<br />{"name":"{form name}",<br />"org":"{user organisation}"}
```
Just for visibility, this morning I mentioned that this was something we picked up when migrating forms-api from grape to rails. We don't actually need a `submission_email` when creating a new form (Only form name and organisation name are required). forms-admin hard codes the empty string here https://github.com/alphagov/forms-admin/blob/main/app/controllers/forms/change_name_controller.rb#L10

File: diagrams/sequence-diagrams/creating-a-form.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  api->>api: Create form
+  api-->>admin: {"id":{form id}}
+  admin-->>browser: 302
+  browser->>admin: GET /forms/{form id}
+  Note right of admin: unclear why this next request happens twice, but it does
+  admin->>api: GET /forms/{form id}
+  admin->>api: GET /forms/{form id}
+  admin->>api: GET /forms/{form id}/pages
```
Comment (aliuk2012) [inline diagrams/sequence-diagrams/creating-a-form.md:28]: For context, the show "Create a form" page has a service that needs to know whether a form has any pages to work out what status to display next the Add/edit question task https://github.com/alphagov/forms-admin/blob/main/app/service/task_status_service.rb#L16-L18

File: diagrams/sequence-diagrams/changing-a-form.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# Changing a form
+
+TODO
```
Comment (aliuk2012) [inline diagrams/sequence-diagrams/changing-a-form.md:3]: Could we leave this file out for the moments? It feels like the current process is already documented as part of the "Set other form details" workflow. I do think its needed but only when we come to do draft/live versioning.  This new draft/live process will be slightly different from the existing "Set other form details". 

I'll branch off this PR in the meantime and try document what draft/live versioning might look like

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+  user->>browser: Visit GOV.UK page<br />that links to the form
+  browser->>govuk: GET
+  govuk-->>browser: Show page
+
+  user->>browser: Click link to form
+  browser->>runner: GET /form/{form id}/{form slug}
+  note over runner: TODO session management
+  runner->>api: GET /forms/{form id}
+  runner->>api: GET /forms/{form id}/pages
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:22]:
```suggestion
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  runner->>runner: determine start page  
+  runner-->>browser: 302
+  browser->>runner: GET /form/{form id}/{form slug}/{start page id}
+  runner->>api: GET /forms/{form id}
+  runner->>api: GET /forms/{form id}/pages
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:27]:
```suggestion
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  browser-->>user: show form
+
+  loop for each form question
+    user->>browser: Provide answer, click "Continue"
+    browser->>runner: POST /form/{form id}/{form slug}/{page id}
+    runner->>api: GET /forms/{form id}
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:33]:
```suggestion
    runner->>api: GET /forms/{form id}/live
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    runner->>api: GET /forms/{form id}/pages
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:34]:
```suggestion
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    runner->>runner: Save user session
+    runner-->>browser: 302
+    browser->>runner: GET /form/{form id}/{form slug}/{next page id}
+    runner->>api: GET /forms/{form id}
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:38]:
```suggestion
    runner->>api: GET /forms/{form id}/live
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    runner->>api: GET /forms/{form id}/pages
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:39]:
```suggestion
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    browser-->>user: Show next question    
+  end
+
+  browser->>runner: GET /form/{form id}/{form slug}/check_your_answers
+  browser-->>user: Show check your answers page
+  user->>browser: Click "Submit"
+  browser->>runner: POST /form/{form id}/{form slug}/submit-answers
+  runner->>api: GET /forms/{form id}
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:47]:
```suggestion
  runner->>api: GET /forms/{form id}/live
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+  runner->>api: GET /forms/{form id}/pages
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:48]:
```suggestion
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+  runner->>notify: client.send_email()
+  runner->>runner: Delete user session
+  runner-->>browser: 302
+  browser->>runner: GET /form/{form id}/{form slug}/submitted
+  runner->>api: GET /forms/{form id}
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:53]:
```suggestion
  runner->>api: GET /forms/{form id}/live
```

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+  runner->>api: GET /forms/{form id}/pages
```
Comment (thomasiles) [inline diagrams/sequence-diagrams/filling-in-a-form.md:54]:
```suggestion
```

File: diagrams/sequence-diagrams/creating-a-form.md (hunk 7 above)
Comment (james-sheppard-gov-uk) [reply diagrams/sequence-diagrams/creating-a-form.md:20]: This PR is to capture what currently happens. Changes to functionality can be proposed in future PRs.

File: diagrams/sequence-diagrams/creating-a-form.md [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+  browser-->>user: show "What is the name of your form?" page
+  user->>browser: Give the form a name<br />click "Save and continue" button
+  browser->>admin: POST /forms/new<br />payload: forms_change_name_form%5Bname%5D={form name}
+  admin->>api: POST /forms<br />{"name":"{form name}",<br />"submission_email":"",<br />"org":"{user orginisation}"}
+  api->>api: Create form
+  api-->>admin: {"id":{form id}}
+  admin-->>browser: 302
+  browser->>admin: GET /forms/{form id}
+  Note right of admin: unclear why this next request happens twice, but it does
+  admin->>api: GET /forms/{form id}
```
[the remark this answers is not in the document]
Comment (james-sheppard-gov-uk) [reply diagrams/sequence-diagrams/creating-a-form.md]: Thanks, added to sequence diagram.

File: diagrams/sequence-diagrams/creating-a-form.md [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+  link admin: GitHub repo @ https://github.com/alphagov/forms-admin
+  participant api as forms-api
+  link api: GitHub repo @ https://github.com/alphagov/forms-api
+
+  note right of user: On the "Create a form" page
+
+  user->>browser: click "Add and edit your questions" link
+  browser->>admin: GET /forms/1/pages/new/type-of-answer
+  Note right of admin: unclear why this next request happens twice, but it does
+  admin->>api: GET /forms/{form id}
```
[the remark this answers is not in the document]
Comment (james-sheppard-gov-uk) [reply diagrams/sequence-diagrams/creating-a-form.md]: Thanks, added to sequence diagram.

File: diagrams/sequence-diagrams/changing-a-form.md (hunk 9 above)
Comment (james-sheppard-gov-uk) [reply diagrams/sequence-diagrams/changing-a-form.md]: I've added something that reflects current (temporary) behaviour. A future PR can show proposed behaviour.

**Comment by james-sheppard-gov-uk** [comment]
> Just wondering if these diagrams would be better on the https://github.com/alphagov/forms-team/wiki? There's no PR flow for the wiki which is either a good thing by reducing the barrier to contributing or a bad thing because its harder to propose a change.

Definitely worth a discussion, but I prefer PR flow for managing proposed changes.

File: diagrams/sequence-diagrams/README.md [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+        user((Person filling in a form))
+        processor((form processor))
+    end
+
+    signon(GOV.UK Signon)
+    govuk(GOV.UK website)
+
+    signon_admin --- signon
+    creator --- signon
+    content --- govuk
```
Comment (DavidBiddle) [inline diagrams/sequence-diagrams/README.md:24]: Should the content designer also be using signon?
```suggestion
    content --- govuk
    content --- signon
```

File: diagrams/sequence-diagrams/authentication.md [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+  actor new_user as new GOV.UK Forms user
+
+  user->>browser: visit GOV.UK Signon Users<br />https://signon.publishing.service.gov.uk/users
+  browser->>signon: GET /users
+  browser-->>user: show list of users
+
+  alt User already exists
+    user->>browser: click on user
+    browser->>signon: GET /users/{user id}/edit
+    user->>browser: check "Has access?", "Update User"
```
Comment (DavidBiddle) [inline diagrams/sequence-diagrams/authentication.md:19]: Suggestion for consistency with the new account route:
```suggestion
    user->>browser: check  GOV.UK Forms "Has access?", "Update User"
```

File: diagrams/sequence-diagrams/README.md (hunk 22 above)
Comment (james-sheppard-gov-uk) [reply diagrams/sequence-diagrams/README.md:24]: They do, but I chose to abstract away how the publishing process works. I don't think there's a benefit in showing the link, plus we'll be moving away from Signon for authentication which will make it even less relevant.

File: diagrams/sequence-diagrams/creating-a-form.md [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+  browser->>admin: GET /forms/{form id}/pages/new
+  browser-->>user: show "Edit question" page
+
+
+  user->>browser: Provide Question text, Hint text (optional), Question settings
+  user->>browser: Click "Save and add next question" OR "Save question" button
+  browser->>admin: POST /forms/{form id}/pages/new
+  admin->>api: POST /forms/{form id}/pages<br />{"question_text":,<br />"hint_text":,<br />"answer_type":,<br />"is_optional":,<br />"answer_settings":{"input_type":}
+  api->>api: create page
+  api-->>admin: { ??? }
```
Comment (DavidBiddle) [inline diagrams/sequence-diagrams/creating-a-form.md:68]: API responds with `201 Created`

```suggestion
  api-->>admin: 201
```

Comment (james-sheppard-gov-uk) [reply diagrams/sequence-diagrams/creating-a-form.md:68]: Thanks @DavidBiddle - I think I was trying to work out what the response json payload looks like, similar to how I had with the POST json payload. Perhaps this level of detail is best left to some API documentation instead so let's go with your suggestion here.

Comment (DavidBiddle) [reply diagrams/sequence-diagrams/creating-a-form.md]: I can help with that too - fromt he code it looks like the payload is just `{"id": 123}` (or whatever the ID of the new page is)

File: diagrams/sequence-diagrams/filling-in-a-form.md [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+  participant inbox as shared email inbox
+  actor processor as Form processor
+
+  user->>browser: Visit GOV.UK page<br />that links to the form
+  browser->>govuk: GET
+  govuk-->>browser: Show page
+
+  user->>browser: Click link to form
+  browser->>runner: GET /form/{form id}/{form slug}
+  note over runner: TODO session management
```
Comment (DavidBiddle) [inline diagrams/sequence-diagrams/filling-in-a-form.md:20]: Is this to do in the current PR, or to do later?

Comment (aliuk2012) [reply diagrams/sequence-diagrams/filling-in-a-form.md:20]: I think we should do this later and improve these diagrams. It seems like a good starting point for us and we can improve as we go along

**Review by aliuk2012 — APPROVED** [review]
Think this is good as a starting point and we can look at improving it further as we progress.

**Review by DavidBiddle — APPROVED** [review]

**Pull request merged**